### PR TITLE
add delete home to the storage provider

### PR DESCRIPTION
**Pull request by C0rby**
Added a delete home method to the storage provider to be able to clean up when removing a user.

### CODE REVIEW

File: cs3/storage/provider/v1beta1/provider_api.proto [hunk 1]
```diff
@@ -724,3 +726,18 @@ message CreateHomeResponse {
   // Opaque information.
   cs3.types.v1beta1.Opaque opaque = 2;
 }
+
+message DeleteHomeRequest {
+  // OPTIONAL.
+  // Opaque information.
+  cs3.types.v1beta1.Opaque opaque = 1;
```
Comment (refs) [inline cs3/storage/provider/v1beta1/provider_api.proto:733]: Having `DeleteHomeRequest` as optional leaks implementation details, for instance, we can call `DeleteHome` call with no arguments and this **COULD** translate into "delete home for the authenticated user", having a side effect.

I'm just saying that we have a chance to make this a pure function and have no side effects by adding a [user](https://cs3org.github.io/cs3apis/#cs3.identity.user.v1beta1.User) to the request 🤷

Comment (C0rby) [reply cs3/storage/provider/v1beta1/provider_api.proto:733]: I would be fine for that. The reason I did it like that is that I wanted the API to be similar to CreateHome.

Comment (butonic) [reply cs3/storage/provider/v1beta1/provider_api.proto:733]: I think CreateHome should also use a user parameter.

**Comment by labkode** [comment]
This API call should be part of an Admin API, we have discussed it in the past and now with this need arising is the right moment to create it. @C0rby can you add this new API here: `cs3/admin/v1beta1/admin_api.proto` with that method?
Note that the method must accept a user ID as argument, this API will be used by site administrators or by robot accounts to automate actions.

**Comment by C0rby** [comment]
Cool.
Just for clarification, should I move the method to the admin_api or copy it to have it in both places?

**Comment by refs** [comment]
> This API call should be part of an Admin API, we have discussed it in the past and now with this need arising is the right moment to create it. @C0rby can you add this new API here: `cs3/admin/v1beta1/admin_api.proto` with that method?
> Note that the method must accept a user ID as argument, this API will be used by site administrators or by robot accounts to automate actions.

thanks for the clarification, I meant `userid`, not an entire user 👯

File: cs3/admin/v1beta1/admin_api.proto [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+  rpc DeprovisionUser(DeprovisionUserRequest) returns (DeprovisionUserResponse);
+}
+
+message ProvisionUserRequest {
+  // OPTIONAL.
+  // Opaque information.
+  cs3.types.v1beta1.Opaque opaque = 1;
+  // REQUIRED.
+  // The user for which the action should be performed.
+  cs3.identity.user.v1beta1.UserId user_id = 2;
```
Comment (labkode) [inline cs3/admin/v1beta1/admin_api.proto:64]: Let's use cs3.identity.user.v1beta1.User instead to give more information to the system when creating an user.

File: cs3/storage/provider/v1beta1/provider_api.proto [hunk 3]
```diff
@@ -724,3 +729,42 @@ message CreateHomeResponse {
...
+message ProvisionStorageRequest {
+  // OPTIONAL.
+  // Opaque information.
+  cs3.types.v1beta1.Opaque opaque = 1;
+  // REQUIRED.
+  // The owner of the storage.
+  cs3.identity.user.v1beta1.UserId owner = 2;
+  // REQUIRED.
+  // The quota for the user.
+  uint64 quota = 3;
```
Comment (labkode) [inline cs3/storage/provider/v1beta1/provider_api.proto:742]: let's redefine this to:
`uint64 quota_max_bytes`
and add also this field
`uint64 quota_max_files`.

The latter is to limit the number of files a user can have, EOS supports that.

File: cs3/storage/provider/v1beta1/provider_api.proto [hunk 4]
```diff
@@ -724,3 +729,42 @@ message CreateHomeResponse {
...
   cs3.types.v1beta1.Opaque opaque = 2;
 }
+
+message ProvisionStorageRequest {
+  // OPTIONAL.
+  // Opaque information.
+  cs3.types.v1beta1.Opaque opaque = 1;
+  // REQUIRED.
+  // The owner of the storage.
+  cs3.identity.user.v1beta1.UserId owner = 2;
```
Comment (labkode) [inline cs3/storage/provider/v1beta1/provider_api.proto:739]: Let's use a full user object here, like in the Admin API, like that the Storage can have more information (like uid/gid) already provided to provision it without having to call the user service.

This function is called only when provisioning users,  the rate of this call will be very low so we can afford to prefetch this info beforehand.

**Review by labkode — CHANGES_REQUESTED** [review]
@C0rby I've added some comments, overall looks pretty good.

**Comment by labkode** [comment]
@C0rby @refs any update on this? Will be nice to merge it in.

**Comment by C0rby** [comment]
I rebased the branch.
From my POV we can merge, right @butonic?

**Comment by kulmann** [comment]
@C0rby could you rebase please? Would be nice to get this in. :-)

**Comment by phil-davis** [comment]
@C0rby are you online this week? Can you rebase this and we can try to push it along. There are things that depend on this that are waiting.

**Comment by phil-davis** [comment]
@labkode @butonic @refs looks like time to review.

There is no response from drone CI - but that will be related to the drone troubles over Christmas...

**Comment by C0rby** [comment]
Well, new year new luck.

Can we merge this? @refs, @butonic, @labkode

**Review by refs — APPROVED** [review]

**Comment by refs** [comment]
@C0rby it LGTM, could we get @ishank011 to have another pair of eyes? :)

**Comment by ishank011** [comment]
@C0rby I think I'm missing something. Why do we have duplicate definitions for (de)provisioning? IMO, the admin API should serve as a gateway to the RPCs in the storage provider, like the gateway API (but we shouldn't add it there so creating a new one makes sense)

**Comment by ishank011** [comment]
@C0rby looks good but how will `ProvisionStorage` be called? I'm guessing from the `ProvisionUser` method in the admin API? How will we supply the quota parameters to it then?

**Comment by C0rby** [comment]
I see your point.
@butonic, what do you say? Add the quota to the ProvisionUser api?

**Comment by butonic** [comment]
As discussed at the CS3 we should pick a more generic name for `CreateHome`. @ishank011 proposed `CreateStorageSpace` which captures the difference between creating a storage vs creating something inside a storage. _Something_ being called a space.

Other options were _CreateTree_, _CreateSpace_, _CreateSubTree_, and _CreateRoot_ but they are either too abstract / graph theory influenced (anything with tree or root) or would invent a new term (space).

_Storage space_ seems to be a good middle ground. Furthermore it can be used not only for user homes but also other _spaces_ eg. project or group _spaces_.

The method needs an owner (which can be a user or a group), optionally quota and opaque data as usual.

cc @ishank011 @C0rby

**Comment by C0rby** [comment]
@jfd, wouldn't CreateStorageSpace also  need a `path` parameter or something like that?

**Comment by butonic** [comment]
CreateStorageSpace where? 

In the gateway? there it would need an indication of which storage should create the space. might be done by path or by storageid.

In a storage provider? No, It is up to the storage provider to determine where to put the new root of the storeage.

In the storage registry? Yes, it needs to know where to mount a storage space. Not only a strorage provitder. But this is a bigger change.

For now, I would limit this change to the StorageProvider Api in the context of the UserProvisioning api: the Gateway can determine the responsible storage provider by calling `storageregistry->GetHome()` and then using a `storageregistry->GetProvider(<home>)`. It can then send the `CreateStorageSpace(<owner>, <quota>)` to it, just as it is currently sending a `CreateHome()`

**Comment by butonic** [comment]
@C0rby  While you are at it could you also add an `UpdateStorageSpace` so we can adjust the quota? We need that for the quota related tests.

**Comment by butonic** [comment]
blocks https://github.com/owncloud/ocis/issues/1290
blocks https://github.com/owncloud/ocis/issues/1313
https://github.com/owncloud/product/issues/247

**Comment by C0rby** [comment]
@ishank011 @labkode, @butonic and I now implemented the storage API how we think it should be.
I removed the admin API again and only added the storage space stuff.

**Review by ishank011 — APPROVED** [review]

**Comment by phil-davis** [comment]
@C0rby is there a PR to `cs3org/reva` that implements this?

**Comment by C0rby** [comment]
@phil-davis, not yet. We still have to do that. It will most likely happen within the next sprint.

**Pull request merged**